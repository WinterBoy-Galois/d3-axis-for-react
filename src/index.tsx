import React from "react";
import type { AxisScale, AxisDomain } from "d3-axis";

function identity(x: any) {
  return x;
}

export enum Orient {
  top = 1,
  right = 2,
  bottom = 3,
  left = 4,
}

function translateX(x: number) {
  return "translate(" + x + ",0)";
}

function translateY(y: number) {
  return "translate(0," + y + ")";
}

export const Axis = <Domain extends AxisDomain>({
  scale,
  ticks,
  tickArguments = [],
  tickValues = null,
  tickFormat = null,
  tickSize,
  tickSizeInner = 6,
  tickSizeOuter = 6,
  tickPadding = 3,
  orient = Orient.bottom,
  offset = typeof window !== "undefined" && window.devicePixelRatio > 1
    ? 0
    : 0.5,
}: {
  scale: AxisScale<Domain>;
  ticks?: any;
  tickArguments?: any[];
  tickValues?: any[] | null;
  // TODO
  tickFormat?: any;
  tickSize?: number;
  tickSizeInner?: number;
  tickSizeOuter?: number;
  tickPadding?: number;
  offset?: number;
  orient?: Orient;
}) => {
  if (tickSize) {
    tickSizeInner = tickSize;
    tickSizeOuter = tickSize;
  }

  if (ticks) {
    tickArguments = ticks;
  }

  function number(scale: AxisScale<Domain>) {
    return (d: any) => {
      const value = scale(d);
      return value === undefined ? 0 : +value;
    };
  }

  function center(scale: AxisScale<Domain>, offset: number) {
    if (scale.bandwidth) {
      offset = Math.max(0, scale.bandwidth() - offset * 2) / 2;
    }
    if ((scale as any).round()) offset = Math.round(offset);
    return (d: Domain) => {
      const value = scale(d);
      return value === undefined ? 0 : value + offset;
    };
  }

  const k = orient === Orient.top || orient === Orient.left ? -1 : 1,
    x = orient === Orient.left || orient === Orient.right ? "x" : "y",
    transform =
      orient === Orient.top || orient === Orient.bottom
        ? translateX
        : translateY;

  // Rendering
  const values =
      tickValues == null
        ? (scale as any).ticks
          ? (scale as any).ticks.apply(scale, tickArguments)
          : scale.domain()
        : tickValues,
    format =
      tickFormat == null
        ? (scale as any).tickFormat
          ? (scale as any).tickFormat.apply(scale, tickArguments)
          : identity
        : tickFormat,
    spacing = Math.max(tickSizeInner, 0) + tickPadding,
    range = scale.range(),
    range0 = +range[0] + offset,
    range1 = +range[range.length - 1] + offset,
    position = (scale.bandwidth ? center : number)(scale.copy(), offset);

  const domainPath =
    orient === Orient.left || orient === Orient.right
      ? tickSizeOuter
        ? "M" +
          k * tickSizeOuter +
          "," +
          range0 +
          "H" +
          offset +
          "V" +
          range1 +
          "H" +
          k * tickSizeOuter
        : "M" + offset + "," + range0 + "V" + range1
      : tickSizeOuter
      ? "M" +
        range0 +
        "," +
        k * tickSizeOuter +
        "V" +
        offset +
        "H" +
        range1 +
        "V" +
        k * tickSizeOuter
      : "M" + range0 + "," + offset + "H" + range1;

  const lineProps = {
    [x + "2"]: k * tickSizeInner,
  };

  const textProps = {
    [x]: k * spacing,
  };

  return (
    <g>
      {values.map((tick: any, i: number) => (
        <g
          className="tick"
          key={i}
          transform={transform(position(tick) + offset)}
        >
          <line stroke="currentColor" {...lineProps} />
          <text
            fill="currentColor"
            {...textProps}
            dy={
              orient === Orient.top
                ? "0em"
                : orient === Orient.bottom
                ? "0.71em"
                : "0.32em"
            }
            fontSize="10"
            fontFamily="sans-serif"
            textAnchor={
              orient === Orient.right
                ? "start"
                : orient === Orient.left
                ? "end"
                : "middle"
            }
          >
            {format(tick)}
          </text>
        </g>
      ))}
      <path
        className="domain"
        stroke="currentColor"
        fill="transparent"
        d={domainPath}
      />
    </g>
  );
};
