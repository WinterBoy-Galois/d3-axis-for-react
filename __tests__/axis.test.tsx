import React from "react";
import renderer from "react-test-renderer";
import { Axis, Orient } from "../src/index";
import { scaleLinear } from "d3";

test("axis with default options", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(<Axis<number> scale={scale} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("axis with top orientation", async () => {
  const scale = scaleLinear().domain([0, 1]).range([0, 1]);
  const component = renderer.create(
    <Axis<number> orient={Orient.top} scale={scale} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
