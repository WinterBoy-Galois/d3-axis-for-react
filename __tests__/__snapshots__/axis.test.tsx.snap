// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`axis with default options 1`] = `
Array [
  <g
    className="tick"
    transform="translate(0.5,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.0
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.6,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.1
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.7,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.2
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.8,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.3
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.9,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.4
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.5
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.1,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.6
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.2,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.7
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.3,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.8
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.4,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.9
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.5,0)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.71em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      1.0
    </text>
  </g>,
  <path
    className="domain"
    d="M0.5,6V0.5H1.5V6"
    fill="transparent"
    stroke="currentColor"
  />,
]
`;

exports[`axis with top orientation 1`] = `
Array [
  <g
    className="tick"
    transform="translate(0.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.0
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.6,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.1
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.7,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.2
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.8,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.3
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0.9,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.4
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.5
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.1,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.6
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.2,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.7
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.3,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.8
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.4,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      0.9
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(1.5,0)"
  >
    <line
      stroke="currentColor"
      y2={-6}
    />
    <text
      dy="0em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={-9}
    >
      1.0
    </text>
  </g>,
  <path
    className="domain"
    d="M0.5,-6V0.5H1.5V-6"
    fill="transparent"
    stroke="currentColor"
  />,
]
`;

exports[`loads and displays greeting 1`] = `
Array [
  <g
    className="tick"
    transform="translate(0,0.5)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.0
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,0.6)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.1
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,0.7)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.2
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,0.8)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.3
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,0.9)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.4
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.5
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1.1)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.6
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1.2)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.7
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1.3)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.8
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1.4)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      0.9
    </text>
  </g>,
  <g
    className="tick"
    transform="translate(0,1.5)"
  >
    <line
      stroke="currentColor"
      y2={6}
    />
    <text
      dy="0.32em"
      fill="currentColor"
      fontFamily="sans-serif"
      fontSize="10"
      textAnchor="middle"
      y={9}
    >
      1.0
    </text>
  </g>,
  <path
    className="domain"
    d="M0.5,6V0.5H1.5V6"
    fill="transparent"
    stroke="currentColor"
  />,
]
`;
